UCR - remove records from datasource where not pass filters

lwyszomi opened the pull request:
@emord @snopoke  I added changes with fix for revert https://github.com/dimagi/commcare-hq/pull/15932

Commit:
UCR - remove records from datasource where not pass filters - with fix to Laboratory adapter

snopoke commented:
Looks ok to me, will let @emord confirm

emord commented on corehq/apps/userreports/laboratory/adapter.py:
```diff
@@ -56,3 +56,10 @@ def best_effort_save(self, doc, eval_context=None):
+        self.sql_adapter.delete(doc)
+
+    def doc_exists(self, doc):
+        return self.es_adapter.doc_exists(doc) or self.sql_adapter.doc_exists(doc)
```
If the doc exists in one but not the other, will one or both of the deletes error?

lwyszomi replied:
@emord Only one will throw the error. I'm checking now if doc exist before deleting it.

Commit:
check if doc exists before delete

emord commented on corehq/apps/userreports/es/adapter.py:
```diff
@@ -207,6 +207,14 @@ def _save_rows(self, rows, doc):
+        return records.count > 0
+
+    def delete(self, doc):
+        if self.doc_exists(doc):
```
This means that doc_exists will be called twice for any deletion. I think instead of checking if it exists again, there should be a try/except block. otherwise you could keep the doc_exists here and just call `delete` in an `else` statement in the pillow.

Commit:
ESAdapter delete - catch KeyNotFound when doc doesn't exist in database

lwyszomi commented:
@emord I added try/catch in ESAdapter and I removed checking if doc exist in db before in SQLAdapter.
Also in SQLAdapter we don't need try/catch because delete method doesn't throw a exception if we want delete document which not exists in db

snopoke commented on corehq/apps/userreports/es/adapter.py:
```diff
@@ -207,6 +207,15 @@ def _save_rows(self, rows, doc):
 
+    def doc_exists(self, doc):
+        records = self.get_query_object().es.doc_id(doc['_id'])
+        return records.count > 0
```
It would be much more efficient to do an `exists` query rather than a count:
```
self.es.exists(self.table_name, 'indicator', doc_id)
```

lwyszomi replied:
@snopoke I changed this.

snopoke commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -123,6 +123,11 @@ def delete(self, doc):
 
+    def doc_exists(self, doc):
+        return self.session_helper.Session.query(
+            self.get_query_object().filter_by(doc_id=doc['_id']).exists()
```
pretty sure this doesn't need to be wrapped in an outer query

lwyszomi replied:
@snopoke I checked this and ```exists()``` doesn't execute the query this only create sub-query which we must execute.

snopoke replied:
Checked the docs and that's correct but from what I can tell this still won't execute they query. You need `.scalar()` to actually make it execute.

snopoke replied:
Perhaps a test case for this function itself would be useful.

lwyszomi replied:
@snopoke Added tests and ```.scalar()```

snopoke commented:
@lwyszomi bump

Commit:
use exists method in es adapter

Commit:
added tests for doc_exists method

snopoke commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -189,6 +189,53 @@ def test_process_doc_from_sql(self, datetime_mock):
+        since = self.pillow.get_change_feed().get_latest_offsets()
+
+        # save case to DB - should also publish to kafka
+        case = _save_sql_case(sample_doc)
```
is it necessary to save the doc here - can't you just do `self.pillow.process_change(doc_to_change(sample_doc))`

That would also save having to cleanup afterwards

snopoke replied:
Same goes for `test_check_if_doc_exist` below

snopoke commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -189,6 +189,53 @@ def test_process_doc_from_sql(self, datetime_mock):
+
+    @override_settings(TESTS_SHOULD_USE_SQL_BACKEND=True)
+    def test_check_if_doc_doesnt_exist(self):
+        self.assertEqual(self.adapter.doc_exists({'_id': "1234"}), False)
```
probably doesn't need to be a separate test - could just run this check before processing the change in the `test_check_if_doc_exist` test.

Also could use `self.assertFalse`

snopoke reviewed (commented):
Looks good, just some minor comments on tests.

Commit:
changes in tests

lwyszomi commented:
@snopoke updated

snopoke commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -189,6 +189,33 @@ def test_process_doc_from_sql(self, datetime_mock):
+        datetime_mock.utcnow.return_value = self.fake_time_now
+        sample_doc, expected_indicators = get_sample_doc_and_indicators(self.fake_time_now)
+
+        self.pillow.process_change(doc_to_change(sample_doc))
```
Should still validate that the doc is in the table after processing:
```
 self._check_sample_doc_state(expected_indicators)
```

lwyszomi replied:
@snopoke Added

snopoke commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -189,6 +189,33 @@ def test_process_doc_from_sql(self, datetime_mock):
+
+        self.pillow.process_change(doc_to_change(sample_doc))
+
+        self.assertEqual(self.adapter.doc_exists(sample_doc), True)
```
nit: `self.assertTrue`

lwyszomi replied:
@snopoke I can't use in this place ```self.assertTrue``` because I want check in this line if the method returns ```True``` and nothing else. For ex: before I added ```scalar()```, method return us an object which means that ```self.assertTrue``` will have positive pass.

emord replied:
@lwyszomi https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertTrue says to use `self.assertIs(expr, True)` for that

lwyszomi replied:
@emord @snopoke updated

Commit:
check doc affter processing

Commit:
replace assertEqual to assertIs
